deployment/ccip/changeset/cs_deploy_token_pools.go

```diff
@@ -0,0 +1,215 @@
+	}
+	newAddresses := deployment.NewMemoryAddressBook()
+
+	state, err := LoadOnchainState(env)
+	if err != nil {
+		return deployment.ChangesetOutput{}, fmt.Errorf("failed to load onchain state: %w", err)
+	}
+
+	for chainSelector, poolConfig := range c.NewPools {
+		chain := env.Chains[chainSelector]
+		chainState := state.Chains[chainSelector]
+
+		_, err := DeployTokenPool(env.Logger, chain, chainState, newAddresses, poolConfig)
+		if err != nil {
+			return deployment.ChangesetOutput{}, fmt.Errorf("failed to deploy %s token pool on %s: %w", c.TokenSymbol, chain.String(), err)
+		}
+	}
+
+	return deployment.ChangesetOutput{
+		AddressBook: newAddresses,
+	}, nil
+}
+
+// deployTokenPool deploys a token pool contract based on a given type & configuration.
+func DeployTokenPool(
+	logger logger.Logger,
+	chain deployment.Chain,
+	chainState CCIPChainState,
+	addressBook deployment.AddressBook,
+	poolConfig DeployTokenPoolInput,
+) (*deployment.ContractDeploy[*token_pool.TokenPool], error) {
+	router := chainState.Router
+	rmnProxy := chainState.RMNProxy
+
+	return deployment.DeployContract(logger, chain, addressBook,
+		func(chain deployment.Chain) deployment.ContractDeploy[*token_pool.TokenPool] {
+			var tpAddr common.Address
+			var tx *types.Transaction
+			var err error
+			switch poolConfig.Type {
+			case BurnMintTokenPool:
+				tpAddr, tx, _, err = burn_mint_token_pool.DeployBurnMintTokenPool(
+					chain.DeployerKey, chain.Client, poolConfig.TokenAddress, poolConfig.LocalTokenDecimals,
+					poolConfig.AllowList, rmnProxy.Address(), router.Address(),
+				)
+			case BurnWithFromMintTokenPool:
+				tpAddr, tx, _, err = burn_with_from_mint_token_pool.DeployBurnWithFromMintTokenPool(
+					chain.DeployerKey, chain.Client, poolConfig.TokenAddress, poolConfig.LocalTokenDecimals,
+					poolConfig.AllowList, rmnProxy.Address(), router.Address(),
+				)
+			case BurnFromMintTokenPool:
+				tpAddr, tx, _, err = burn_from_mint_token_pool.DeployBurnFromMintTokenPool(
+					chain.DeployerKey, chain.Client, poolConfig.TokenAddress, poolConfig.LocalTokenDecimals,
+					poolConfig.AllowList, rmnProxy.Address(), router.Address(),
+				)
+			case LockReleaseTokenPool:
+				tpAddr, tx, _, err = lock_release_token_pool.DeployLockReleaseTokenPool(
+					chain.DeployerKey, chain.Client, poolConfig.TokenAddress, poolConfig.LocalTokenDecimals,
+					poolConfig.AllowList, rmnProxy.Address(), *poolConfig.AcceptLiquidity, router.Address(),
+				)
+			}
+			var tp *token_pool.TokenPool
+			if err == nil { // prevents overwriting the error (also, if there were an error with deployment, converting to an abstract token pool wouldn't be useful)
+				tp, err = token_pool.NewTokenPool(tpAddr, chain.Client)
+			}
+			return deployment.ContractDeploy[*token_pool.TokenPool]{
+				Address: tpAddr, Contract: tp, Tv: deployment.NewTypeAndVersion(poolConfig.Type, currentTokenPoolVersion), Tx: tx, Err: err,
```

Comment:
Style nit: can we split this into separate lines?

Reply:
https://github.com/smartcontractkit/chainlink/pull/15877/commits/afc13ca92a4c18ced03712680bcab6df42c0f512